[Convergence] Extend cycles to include outside uses of tokens

ssahasra opened the pull request:
When a convergence control token T defined at an operation D in a cycle
C that is used by an operation U outside C, the cycle is said to be
extended up to U. This because the use of the convergence control T
requires that two threads that execute U must execute converged dynamic
instances of U if and only if they previously executed converged dynamic
instances of D.

For more information including a high-level C-like example, see
https://llvm.org/docs//ConvergentOperations.html

This change introduces a pass that captures this token semantics by
literally extending the cycle C to include every path from C to U.

ssahasra committed:
[Convergence] Extend cycles to include token uses outside

Whe a convergence control token T defined at an operation D in a cycle C that is
used by an operation U outside C, the cycle is said to be extended up to U. This
because the use of the convergence control T requires that two threads that
execute U must execute converged dynamic instances of U if and only if they
previously executed converged dynamic instances of D.

For more information including a high-level C-like example, see
https://llvm.org/docs//ConvergentOperations.html

This change introduces a pass that captures this token semantics by literally
extending the cycle C to include every path from C to U.

jayfoad commented:
> When a convergence control token T defined at an operation D in a cycle
> C that is used by an operation U outside C, the cycle is said to be
> extended up to U.

Nit: remove "that" in the second line?

jayfoad commented:
Can you explain a bit more about the big picture?

It looks like the new pass modifies the IR by adding some dummy branches which a pass like jump threading could easily remove again. So, after running the new pass, can you really rely on the IR remaining in "extended cycle" form for very long?

What effect do the dummy branches have on final codegen?

Why implement this by modifying the IR, instead of providing an analysis that reports the extended form of cycles on unmodified IR?

ssahasra committed:
rebase and squash incremental updates

- Actually update the DomTree.
- Declare that DomTree and CycleInfo are preserved.
- Additional explanatory note at the top of the file.
- use depth_first to visit blocks in a stable way
- clang format
- move the pass to Transforms/Utils

ssahasra commented:
> Can you explain a bit more about the big picture?

Sure. I have also added a new note to the cpp, paraphrasing most of the answers.

> It looks like the new pass modifies the IR by adding some dummy branches which a pass like jump threading could easily remove again. So, after running the new pass, can you really rely on the IR remaining in "extended cycle" form for very long?

It's not clear yet whether this extended cycle can be preserved, and also whether existing transforms should be modified to preserve it if they can. I would like to believe that all transforms should be made aware of token semantics, but that by itself is not sufficient to answer this question.

> What effect do the dummy branches have on final codegen?

That's what we want to find out. This transform will first be used just before the structurizer, so that we can implement tokens end-to-end on AMDGPU. We may eventually use it with the wave transform, or possibly incorporate it into the wave transform.

> Why implement this by modifying the IR, instead of providing an analysis that reports the extended form of cycles on unmodified IR?

Because CodeGen. Also, in the big picture, an analysis will be the right choice if we decide to teach every transform about extended cycles. Otherwise, encoding the cycle extension into the CFG itself (provided it can or should be preserved) allows all the transforms (such as loop unrolling) to just continue doing what they do.

ssahasra commented:
> It looks like the new pass modifies the IR by adding some dummy branches which a pass like jump threading could easily remove again. So, after running the new pass, can you really rely on the IR remaining in "extended cycle" form for very long?

#100185 provides a very useful insight. Extended cycles matter only if an exiting branch is divergent, _and_ a token defined inside the cycle is used outside of it. While that PR is working around divergent branches for performance gain, extended cycles will have to disable the same optimizations for correctness.

ssahasra committed:
rename the pass to put verb first
